sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs

```diff
@@ -198,58 +239,331 @@ public async Task StopAsync()
+                // maximum owned partition count should be targeted.
+
+                var maximumOwnedPartitionsCount = minimumOwnedPartitionsCount + 1;
+
+                var stealablePartitions = activeOwnership
+                    .Where(ownership => partitionDistribution[ownership.OwnerIdentifier] > maximumOwnedPartitionsCount)
+                    .Select(ownership => ownership.PartitionId)
+                    .ToList();
+
+                // Here's the important part.  If there are no processors that have exceeded the maximum owned partition count allowed, we may
+                // need to steal from the processors that have exactly the maximum amount.  If this instance is below the minimum count, then
+                // we have no choice as we need to enforce balancing.  Otherwise, leave it as it is because the distribution wouldn't change.
+
+                if (!stealablePartitions.Any() && ownedPartitionsCount < minimumOwnedPartitionsCount)
+                {
+                    stealablePartitions = activeOwnership
+                        .Where(ownership => partitionDistribution[ownership.OwnerIdentifier] == maximumOwnedPartitionsCount)
+                        .Select(ownership => ownership.PartitionId)
+                        .ToList();
+                }
+
+                // If any stealable partitions were found, randomly pick one of them to claim.
+
+                if (stealablePartitions.Any())
+                {
+                    var index = s_random.Value.Next(stealablePartitions.Count);
+
+                    return await ClaimOwnershipAsync(stealablePartitions[index], completeOwnershipList).ConfigureAwait(false);
+                }
+            }
+
+            // No ownership was claimed.
+
+            return null;
+        }
+
+        /// <summary>
+        ///   Creates a new partition pump associated with the specified partition.  Partition pumps that are overwritten by the creation
+        ///   of a new one are properly stopped.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The identifier of the Event Hub partition the partition pump will be associated with.  Events will be read only from this partition.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task AddOrOverwritePartitionPumpAsync(string partitionId)
+        {
+            // Remove and stop the existing partition pump if it exists.  We are not specifying any close reason because partition
+            // pumps only are overwritten in case of failure.  In these cases, the close reason is delegated to the pump as it may
+            // have more information about what caused the failure.
+
+            await RemovePartitionPumpIfItExistsAsync(partitionId).ConfigureAwait(false);
+
+            // Create the new partition pump and add it to the dictionary.
```

Comment:
nit:  Create and start.   You're not capturing ones that you've created and failed to start.   (which I think is the correct way to go)